c/src/lib/libbsp/arm/shared/stm32fxxxx/spi/hal-spi.c

```diff
@@ -0,0 +1,478 @@
+    rtems_set_errno_and_return_minus_one( ETIMEDOUT );
+  }
+}
+
+/**
+ * Initalize GPIO Pins
+ * Note: If software based slave select, gpio pin for the slave select pin
+ *       has to be configured and initialized by the application.
+ */
+static void stm32_spi_gpio_init( SPI_Instance spi_instance )
+{
+  GPIO_InitTypeDef GPIO_InitStruct;
+
+  /* Initalize SPI Clock */
+  stm32_spi_initialize_spi_clock( spi_instance );
+
+  /* Initalize GPIO Clock */
+  stm32_spi_initialize_gpio_clock( spi_instance );
+
+  GPIO_InitStruct.Mode = GPIO_MODE_AF_PP;
+  GPIO_InitStruct.Pull = GPIO_PULLDOWN;
+  GPIO_InitStruct.Speed = GPIO_SPEED_FAST;
+  GPIO_InitStruct.Alternate = spi_config[ spi_instance ].alternate_func;
+
+  /* SPI SCK Pin */
+  GPIO_InitStruct.Pin = spi_config[ spi_instance ].sck_pin;
+  HAL_GPIO_Init( spi_config[ spi_instance ].port, &GPIO_InitStruct );
+
+  /* SPI MOSI Pin */
+  GPIO_InitStruct.Pin = spi_config[ spi_instance ].mosi_pin;
+  HAL_GPIO_Init( spi_config[ spi_instance ].port, &GPIO_InitStruct );
+
+  /* SPI MISO Pin */
+  GPIO_InitStruct.Pin = spi_config[ spi_instance ].miso_pin;
+  HAL_GPIO_Init( spi_config[ spi_instance ].port, &GPIO_InitStruct );
+}
+
+/**
+ * SPI Initalize Instance
+ */
+static int stm32_spi_init( stm32_spi_bus *bus )
+{
+  rtems_status_code sc;
+
+  _Assert( bus != NULL );
+
+  /* Initialize GPIO pins for SPI */
+  stm32_spi_gpio_init( bus->instance );
+
+  /* Initialize SPI */
+  /* Set the SPI parameters */
+  bus->handle.Instance = spi_config[ bus->instance ].instance;
+
+  bus->handle.Init.BaudRatePrescaler = SPI_BAUDRATEPRESCALER_2;
+  bus->handle.Init.Direction = SPI_DIRECTION_1LINE;
+  bus->handle.Init.CLKPhase = SPI_PHASE_1EDGE;
+  bus->handle.Init.CLKPolarity = SPI_POLARITY_LOW;
+  bus->handle.Init.CRCCalculation = SPI_CRCCALCULATION_DISABLE;
+  bus->handle.Init.CRCPolynomial = 7;
+  bus->handle.Init.DataSize = SPI_DATASIZE_16BIT;
+  bus->handle.Init.FirstBit = SPI_FIRSTBIT_MSB;
+  bus->handle.Init.NSS = SPI_NSS_SOFT;
+  bus->handle.Init.TIMode = SPI_TIMODE_DISABLE;
+
+  bus->handle.Init.Mode = SPI_MODE_MASTER;
+
+  if ( HAL_SPI_Init( &bus->handle ) != HAL_OK ) {
+    ( *bus->base.destroy )( &bus->base );
+    rtems_set_errno_and_return_minus_one( EIO );
+  }
+
+  /* Install SPI vectors */
+  sc = rtems_interrupt_handler_install(
+    spi_config[ bus->instance ].vector_num,
+    "SPI Event Interrupt",
+    RTEMS_INTERRUPT_UNIQUE,
+    stm32_spi_event_irq,
+    bus
+       );
+
+  if ( sc != RTEMS_SUCCESSFUL ) {
+    ( *bus->base.destroy )( &bus->base );
+    rtems_set_errno_and_return_minus_one( EIO );
+  }
+
+  return 0;
+}
+
+/**
+ * spi Bus Destroy and De-initialise
+ */
+static int stm32_spi_deinit_destroy( stm32_spi_bus *bus )
+{
+  rtems_status_code sc;
+
+  _Assert( bus != NULL );
+
+  sc = rtems_interrupt_handler_remove( spi_config[ bus->instance ].vector_num,
+    stm32_spi_event_irq,
+    bus );
+  _Assert( sc == RTEMS_SUCCESSFUL );
+  (void) sc;
+
+  spi_bus_destroy_and_free( &bus->base );
```

Comment:
Bus destroy shouldn't be here in the same function (which is called by close() ). This won't allow open()'ing of the bus after it is close()'ed. Declare a separate function - stm32_spi_deinit_destroy() and destroy here.